### PR TITLE
fix: find correct retry node when using `templateRef`. Fixes: #12633

### PR DESCRIPTION
**Pull request by shuangkun**
The root cause is pod have `retryStrategy`, the boundaryNode didn't have template (use `templateRef`), so when find retryNode,  probabilistically found others and cause `"failed to get a template"`.

Some info help review:
Failed workflow(find the wrong retry node):
image
Succeed worklfow:
image

Fixes: #12633

### CODE REVIEW

File: workflow/controller/retry_tweak.go [hunk 1]
```diff
@@ -15,12 +15,23 @@ type RetryTweak = func(retryStrategy wfv1.RetryStrategy, nodes wfv1.Nodes, pod *
...
+		for _, node := range nodes {
+			if node.Type == wfv1.NodeTypeRetry && node.TemplateName == templateName {
+				return &node
+			}
 		}
 	}
+	if boundaryNode.TemplateRef != nil {
+		templateRef := boundaryNode.TemplateRef
+		for _, node := range nodes {
+			if node.Type == wfv1.NodeTypeRetry && node.TemplateRef != nil && node.TemplateRef.Name == templateRef.Name && node.TemplateRef.Template == templateRef.Template {
```
Comment (Joibel) [inline workflow/controller/retry_tweak.go:29]: Why do we check a match of both `TemplateRef.Name` and `TemplateRef.Template` here?

Comment (Joibel) [reply workflow/controller/retry_tweak.go:29]: No, you're right. I was looking in the wrong place at this.

**Review by Joibel — APPROVED** [review]
LGTM, thanks for digging down to this.

File: workflow/controller/retry_tweak_test.go [hunk 2]
```diff
@@ -59,6 +59,24 @@ func TestFindRetryNode(t *testing.T) {
...
 			TemplateName: "tmpl2",
 		},
+		"E1": wfv1.NodeStatus{
+			ID:         "E1",
+			Type:       wfv1.NodeTypeRetry,
+			Phase:      wfv1.NodeRunning,
+			BoundaryID: "A1",
+			Children:   []string{},
+			TemplateRef: &wfv1.TemplateRef{
+				Name: "tmpl1",
```
Comment (agilgur5) [inline workflow/controller/retry_tweak_test.go:69]: shouldn't this have a `template` field as well? the test is passing with only `name`?

Comment (shuangkun) [reply workflow/controller/retry_tweak_test.go]: Ok,I have added it.

Comment (agilgur5) [reply workflow/controller/retry_tweak_test.go]: > the test is passing with only `name`?

Oh I guess the test passed bc `nil == nil`. The code checks whether the `templateRef` struct isn't `nil`, but not the individual fields.
They should never be `nil` in a real use-case, so I suppose that is a test edge-case

**Review by agilgur5 — APPROVED** [review]
LGTM, thanks for root causing and fixing this!
